File: pc/McController.Server/Input/CameraCurve.cs
using McController.Server.Config;

namespace McController.Server.Input;

/// <summary>
/// Two-layer camera transform applied to raw look deltas before SendInput.
///
/// Layer 1 — Developer curve (hidden from end users in production):
///   Speed-dependent acceleration. Lets quick swipes turn the view fast
///   while slow drags stay precise for fine aim.
///
/// Layer 2 — User sensitivity:
///   Single 0.5..3.0 multiplier exposed in the user-facing UI.
///
/// A residual fractional component is carried across calls so that when
/// sensitivity is low, sequences of small deltas eventually accumulate
/// to a 1-pixel output instead of being truncated to zero.
/// </summary>
public sealed class CameraCurve
{
    private readonly ServerConfig _config;
    private float _residualX;
    private float _residualY;

    public CameraCurve(ServerConfig config)
    {
        _config = config;
    }

    public (int sdx, int sdy) Apply(int rawDx, int rawDy)
    {
        var cam = _config.Camera;
        var speed = MathF.Sqrt((float)(rawDx * rawDx + rawDy * rawDy));

        float accelMul = cam.CurveType switch
        {
            CurveType.Power => Math.Min(
                1f + cam.AccelFactor * MathF.Pow(speed, cam.AccelExp),
                cam.MaxAccelMultiplier),
            _ => 1f,
        };

        var scale = cam.UserSensitivity * accelMul;
        var fx = rawDx * scale + _residualX;
        var fy = rawDy * scale + _residualY;
        var ix = (int)MathF.Truncate(fx);
        var iy = (int)MathF.Truncate(fy);
        _residualX = fx - ix;
        _residualY = fy - iy;
        return (ix, iy);
    }

    public void Reset()
    {
        _residualX = 0;
        _residualY = 0;
    }
}

File: mac/McController/Core/Input/CameraCurve.swift
import Foundation

/// Two-layer camera transform applied to raw look deltas before injection.
/// Mirrors `CameraCurve.cs` precisely — the math is identical, so a
/// profile authored on Windows feels the same after carrying over.
///
/// Layer 1 — Developer curve (hidden from end users in production):
///   Speed-dependent acceleration. Quick swipes turn fast, slow drags
///   stay precise for fine aim.
///
/// Layer 2 — User sensitivity:
///   Single 0.5..3.0 multiplier exposed in the user-facing UI.
///
/// Carries a residual fractional component across calls so a stream of
/// sub-pixel deltas eventually accumulates to a 1-pixel output instead
/// of being truncated to zero every time.
final class CameraCurve {

    private let config: ServerConfig
    private var residualX: Float = 0
    private var residualY: Float = 0

    init(config: ServerConfig) {
        self.config = config
    }

    func apply(rawDx: Float, rawDy: Float) -> (Int, Int) {
        let cam = config.camera
        let speed = (rawDx * rawDx + rawDy * rawDy).squareRoot()

        let accelMul: Float
        switch cam.curveType {
        case .power:
            let raw = 1 + cam.accelFactor * pow(speed, cam.accelExp)
            accelMul = min(raw, cam.maxAccelMultiplier)
        case .linear:
            accelMul = 1
        }

        let scale = cam.userSensitivity * accelMul
        let fx = rawDx * scale + residualX
        let fy = rawDy * scale + residualY
        let ix = Int(fx.rounded(.towardZero))
        let iy = Int(fy.rounded(.towardZero))
        residualX = fx - Float(ix)
        residualY = fy - Float(iy)
        return (ix, iy)
    }

    func reset() {
        residualX = 0
        residualY = 0
    }
}
